ui/webui/src/InstallationLanguage.jsx

```diff
@@ -39,105 +37,112 @@ const getLanguageNativeName = lang => lang["native-name"].v;
 const getLocaleId = locale => locale["locale-id"].v;
 const getLocaleNativeName = locale => locale["native-name"].v;
 
-const LanguageSelector = ({ lang, onSelectLang }) => {
-    const [isOpen, setIsOpen] = useState();
-    const [languages, setLanguages] = useState([]);
-    const [locales, setLocales] = useState({});
-    const [selectedItem, setSelectedItem] = useState();
-    const address = useContext(AddressContext);
-
-    const localizationProxy = useObject(() => {
-        const client = cockpit.dbus("org.fedoraproject.Anaconda.Modules.Localization", { superuser: "try", bus: "none", address });
-        const proxy = client.proxy(
+class LanguageSelector extends React.Component {
+    constructor (props) {
+        super(props);
+        this.state = {
+            languages: [],
+            locales: [],
+        };
+        this.updateDefaultSelection = this.updateDefaultSelection.bind(this);
+    }
+
+    componentDidMount () {
+        const client = cockpit.dbus(
             "org.fedoraproject.Anaconda.Modules.Localization",
-            "/org/fedoraproject/Anaconda/Modules/Localization",
+            { superuser: "try", bus: "none", address: this.context }
         );
+        const call = (method, args) => client.call(
+            "/org/fedoraproject/Anaconda/Modules/Localization",
+            "org.fedoraproject.Anaconda.Modules.Localization",
+            method, args
+        ).then(res => Promise.resolve(res[0]));
 
-        return proxy;
-    }, null, [address]);
-
-    useEvent(localizationProxy, "changed", (event, data) => {
-        localizationProxy.GetLanguages().then(languages => {
+        call("GetLanguages", []).then(ret => {
+            const languages = ret;
             // Create the languages state object
-            Promise.all(languages.map(lang => localizationProxy.GetLanguageData(lang))).then(setLanguages);
+            Promise.all(languages.map(lang => call("GetLanguageData", [lang])))
+                    .then(langs => this.setState({ languages: langs }));
 
             // Create the locales state object
-            Promise.all(languages.map(lang => localizationProxy.GetLocales(lang))).then(res => {
-                return Promise.all(
-                    res.map((langLocales) => {
-                        return Promise.all(langLocales.map(locale => localizationProxy.GetLocaleData(locale)));
+            Promise.all(languages.map(lang => call("GetLocales", [lang])))
+                    .then(res => {
+                        return Promise.all(
+                            res.map(langLocales => {
+                                return Promise.all(langLocales.map(locale =>
+                                    call("GetLocaleData", [locale])
```

Comment:
This looks a little crazy I think. :) I wonder if there is perhaps some way to run this in a worker thread or something, defined as a plain sequence of commands and just return the result to a callback ?

Reply:
I don't think it can be turned into anything sequential. Once you start `Promise`-ing you can never stop, you have to chain it all together. And of course D-Bus "calls" here are promises, because they are in reality XHR to cockpit that does the actual call (I think!).

What could be done is reformat the whitespace, or use less anonymous functions, or use more intermediate variables, maybe. I think all of these would help somewhat. You could also get an editor that shows vertical lines for indentation levels and highlights matching braces etc. (except not on github of course). You could also reduce indentation to 2 spaces. Unless the linter exterminates you for trying to think about any of these and losing the spirit of modern javascript?

I guess the overall point is that some time in the mid-future, we might want to sit down and rethink how the code should be formatted. I can't really read this either, especially if there's a review chat in the middle of it.

Reply:
I am working now on moving the API calls as functions to seperate files, and I think that will massively increase readability.